### PR TITLE
Components of VAE

### LINKED ISSUE #170: Implement Variational AutoEncoder
Reference : http://arxiv.org/abs/1312.6114

### PR DESCRIPTION
**Pull request by unnonouno**
- KL-divergence between a given gaussian and the normal gaussian
- NLL of Bernoulli distribution
- NLL of Gaussian distribution

fix #170

### CODE REVIEW

File: chainer/functions/vae.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import math
+
+import chainer.functions as F
+from chainer import variable
+
+
+def gaussian_kl_divergence(z_mean, z_ln_var):
+    """Calculate KL-divergence between a given gaussian and the normal one.
```
Comment (delta2323) [inline chainer/functions/vae.py]: Normal distribution is an alias of Gaussian distribution. So we should call `N(0, 1)` standard normal distribution

File: chainer/functions/vae.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    Given two variable ``z_mean`` representing :math:`\\mu` and ``z_ln_var``
+    representing :math:`\\log(\\sigma^2)`, this function returns a variable
+    representing KL-divergence between :math:`N(\\mu, \\sigma^2)` and
+    the normal Gaussian :math:`N(0, 1)`.
+
+    Args:
+        z_mean (~chainer.Variable): A variable representing mean of the given
+            gaussian distribution.
+        z_ln_var (~chainer.Variable): A variable representing logarithm of
+            variance of the given gaussian distribution.
```
Comment (delta2323) [inline chainer/functions/vae.py]: I think we do not need *the* here.

File: chainer/functions/vae.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    Returns:
+        ~chainer.Variable: A variable representing KL-divergence between the
```
Comment (delta2323) [inline chainer/functions/vae.py]: ditto(the)

File: chainer/functions/vae.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            given gaussian distribution and the normal gaussian.
```
Comment (delta2323) [inline chainer/functions/vae.py]: ditto(normal gaussian)

File: chainer/functions/vae.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+from chainer import variable
+
+
+def gaussian_kl_divergence(z_mean, z_ln_var):
+    """Calculate KL-divergence between a given gaussian and the normal one.
+
+    Given two variable ``z_mean`` representing :math:`\\mu` and ``z_ln_var``
+    representing :math:`\\log(\\sigma^2)`, this function returns a variable
+    representing KL-divergence between :math:`N(\\mu, \\sigma^2)` and
+    the normal Gaussian :math:`N(0, 1)`.
```
Comment (delta2323) [inline chainer/functions/vae.py]: We should note that `N(mu, sigma^2)` and `N(0, 1)` is multi-dimensional. In concrete, we should add that `mu` and sigma is vectors. We also write `N(mu, sigma^2)` and `N(0, 1)` as `N(mu, sigma^2 I)` and `N(0, I)` where I is a identity matrix.

File: chainer/functions/vae.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    .. math::
+
+        -B(x; p) = -\\sum_i {x_i \\log(p_i) + (1 - x_i)\\log(1 - p_i)},
+
+    where :math:`p = \\sigma(y)`, and :math:`\\sigma(\\cdot)` is a sigmoid
+    funciton.
+
+    When :math:`x \\in \\{0, 1\\}`, the return value is equal to negative
+    log-likelihoood on a Bernoulli distribution.
```
Comment (delta2323) [inline chainer/functions/vae.py]: typo: likelihood

Comment (delta2323) [inline chainer/functions/vae.py]: Does this sentence say same thing as the first paragraph of the document?

File: chainer/functions/vae.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+    .. note::
+
+       As this funtion uses a sigmoid function, you can pass a result of
+       full-connect layer (that means :class:`Linear`) to this function
```
Comment (delta2323) [inline chainer/functions/vae.py]: I think fully-connected layer is more common.

File: chainer/functions/vae.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    """Calculate KL-divergence between a given gaussian and the normal one.
+
+    Given two variable ``z_mean`` representing :math:`\\mu` and ``z_ln_var``
+    representing :math:`\\log(\\sigma^2)`, this function returns a variable
+    representing KL-divergence between :math:`N(\\mu, \\sigma^2)` and
+    the normal Gaussian :math:`N(0, 1)`.
+
+    Args:
+        z_mean (~chainer.Variable): A variable representing mean of the given
+            gaussian distribution.
```
Comment (delta2323) [inline chainer/functions/vae.py]: It is better to add explanation that `z_mean` indicates `mu` as argument list of `gaussian_nll` does.

File: chainer/functions/vae.py (hunk 2 above)
Comment (delta2323) [inline chainer/functions/vae.py]: ditto(add `log(sigma^2)`)

File: tests/functions_tests/test_vae.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    cuda.init()
+
+
+class TestGaussianKLDivergence(unittest.TestCase):
+
+    def setUp(self):
+        self.mean = numpy.random.uniform(-1, 1, (3,)).astype(numpy.float32)
+        self.ln_var = numpy.random.uniform(-1, 1, (3,)).astype(numpy.float32)
+
+        # Refer to Appendix B in the original paper
```
Comment (delta2323) [inline tests/functions_tests/test_vae.py]: Please add reference to the original paper

File: chainer/functions/vae.py (hunk 5 above)
Comment (delta2323) [inline chainer/functions/vae.py]: As KL-divergence is asymmetric (i.e. KL(p||q) != KL(q||p)), we should specify which is first term and which is second.
I think "KL-divergence of N(mu, sigma^2) from N(0, 1)" is enough.

File: chainer/functions/vae.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+    It returns negative log-likelihood of :math:`x` on a Gaussian distribution
+    :math:`N(\\mu, \\sigma^2)`, where :math:`\\mu` is given as ``mean`` and
+    :math:`\\log(\\sigma^2)` is given as ``ln_var``.
+
+    .. math::
+
+        -\\log N(x; \\mu, \\sigma^2) =
+        \\log((\\sqrt{2\\pi})^D \\sqrt{|S|}) +
+        \\frac{1}{2}(x - \\mu)S^{-1}(x - \\mu)
```
Comment (delta2323) [inline chainer/functions/vae.py]: First `(x-mu)` should be transposed (if we use column vectors).

File: chainer/functions/vae.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    """Calculate negative log-likelihood of Gaussian distribution.
+
+    It returns negative log-likelihood of :math:`x` on a Gaussian distribution
+    :math:`N(\\mu, \\sigma^2)`, where :math:`\\mu` is given as ``mean`` and
+    :math:`\\log(\\sigma^2)` is given as ``ln_var``.
+
+    .. math::
+
+        -\\log N(x; \\mu, \\sigma^2) =
+        \\log((\\sqrt{2\\pi})^D \\sqrt{|S|}) +
```
Comment (delta2323) [inline chainer/functions/vae.py]: It is just a matter of taste but I like `¥log(¥sort{(2¥pi)^D |S|})` better.

File: chainer/functions/vae.py (hunk 1 above)
Comment (unnonouno) [inline chainer/functions/vae.py]: woops

File: chainer/functions/vae.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
+import math
+
+import chainer.functions as F
+from chainer import variable
+
+
+def gaussian_kl_divergence(z_mean, z_ln_var):
```
Comment (delta2323) [inline chainer/functions/vae.py]: If we separate this function from the context of VAE, users of this function can be confused what `z` in `z_mean` and `z_ln_var` stands for. So `mean` and `ln_var` is enough as names of arguments.

File: chainer/functions/vae.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    assert isinstance(y, variable.Variable)
+
+    return F.sum(F.softplus(-y)) + F.sum(y) - F.sum(y * x)
+
+
+def gaussian_nll(x, mean, ln_var):
+    """Calculate negative log-likelihood of Gaussian distribution.
+
+    It returns negative log-likelihood of :math:`x` on a Gaussian distribution
+    :math:`N(\\mu, \\sigma^2)`, where :math:`\\mu` is given as ``mean`` and
```
Comment (delta2323) [inline chainer/functions/vae.py]: `N(mu, S)` is more appropriate (as done in `gaussian_kl_divergence`)

**Comment by delta2323** [comment]
LGTM :+1:

**Pull request merged**